features/FxA-83-sign-in-confirmation-email/README.md

```diff
@@ -0,0 +1,290 @@
+
+Add a notion of verification
+to sessionTokens and keyFetchTokens.
+Tokens that are unverified
+have reduced powers (see below).
+Token verification is achieved
+by following a link sent by email.
+
+### What happens when a Sync user signs in?
+
+1. User submits form.
+
+2. Content server requests `POST /account/login?keys=true`.
+
+3. Auth server generates `tokenVerificationId`
+   and sends it with `tokenVerified:false`
+   in request to `PUT /sessionToken/:tokenId`.
+
+4. Auth server sends same `tokenVerificationId`
+   and `tokenVerified:false`
+   in request to `PUT /keyFetchToken/:tokenId`.
+
+5. Auth server invokes the mailer
+   to send a confirmation email,
+   which includes `tokenVerificationId` in the URL.
+
+6. Auth server sends response back to content server,
+   including in the data:
+   ```
+   {
+     "sessionToken": sessionTokenId,
+     "keyFetchToken": keyFetchTokenId,
+     "challengeReason": "signin",
+     "challengeMethod": "email"
+   }
+   ```
+
+7. Content server sends web channel message
+   notifying the browser of successful sign-in.
+
+8. Content server navigates to confirm-signin view.
+
+9. Content server starts polling
+   on `GET /token/:tokenId/status`.
+
+10. User clicks link in confirmation email.
+
+11. Content server requests `POST /token/verify`.
+
+12. Auth server requests `POST /token/:tokenVerificationId/verify`.
+    Both tokens are verified in the database.
+	If the account is unverified,
+	that also gets verified here.
+
+13. Concurrently:
+
+    * Auth server responds to verification request,
+      including verified tokens in the data.
+
+    * Auth server changes polling request responses,
+      including verified tokens in the data.
+
+14. Content server navigates to about:accounts.
+
+### What happens when an OAuth user signs in?
+
+1. User submits form.
+
+2. Content server requests `POST /account/login`.
+
+3. Auth server generates `tokenVerificationId`
+   and sends it with `tokenVerified:false`
+   in request to `PUT /sessionToken/:tokenId`.
+
+4. Auth server sends same `tokenVerificationId`
+   and `tokenVerified:false`
+   in request to `PUT /keyFetchToken/:tokenId`.
+
+5. Auth server sends response back to content server.
+
+6. Content server redirects to OAuth relier.
+
+### What happens when an unverified client requests `/certificate/sign`?
+
+If the sessionToken is unverified,
+the request will succeed.
+However, the verification status is stored
+in the certificate as `fxa-tokenVerified`,
+alongside other `fxa-` properties.
+The Sync server can subsequently reject assertions
+where `fxa-tokenVerified` is `false`.
+
+### What happens when an unverified client requests `/account/keys`?
+
+If the keyFetchToken is unverified,
+the request will fail with errno 102.
+
+### How does this affect users on legacy clients?
+
+They will receive the confirmation email,
+but they will not see the sign-in confirmation screen.
+Subsequent operations may fail with token errors
```

Comment:
legacy users will attempt to sign in by calling `/account/login` on the auth server. The auth server will generate an unverified sessionToken and respond with `verified: false`. The client will redirect the user to their already existing "confirm your email" screen where the client will start polling `/recovery_email/status` using the sessionToken returned by `/account/login`. `/recovery_email/status` will return `verified: false` because the sessionToken is unverified. When the user verified their email, `/recovery_email/status` will return `verified: true` and the legacy client knows the Sync sign in can continue.